pkg/gardenadm/translator/static_pod.go

```diff
@@ -0,0 +1,140 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package translator
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"slices"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+// Translate translates the given object into a list of files containing static pod manifests as well as ConfigMaps and
+// Secrets that can be injected into an OperatingSystemConfig.
+func Translate(ctx context.Context, c client.Client, o client.Object) ([]extensionsv1alpha1.File, error) {
+	switch obj := o.(type) {
+	case *appsv1.Deployment:
+		return translatePodTemplate(ctx, c, obj.ObjectMeta, obj.Spec.Template)
+	// TODO(rfranzke): Consider adding support for StatefulSet in the future.
+	default:
+		return nil, fmt.Errorf("unsupported object type %T", o)
+	}
+}
+
+func translatePodTemplate(ctx context.Context, c client.Client, objectMeta metav1.ObjectMeta, podTemplate corev1.PodTemplateSpec) ([]extensionsv1alpha1.File, error) {
+	pod := &corev1.Pod{ObjectMeta: podTemplate.ObjectMeta, Spec: podTemplate.Spec}
+	pod.Name = objectMeta.Name
+	pod.Namespace = objectMeta.Namespace
+
+	translateSpec(&pod.Spec)
+
+	filesFromVolumes, err := translateVolumes(ctx, c, pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed translating volumes for static pod %s: %w", client.ObjectKeyFromObject(pod), err)
+	}
+
+	staticPodYAML, err := kubernetesutils.Serialize(pod, c.Scheme())
+	if err != nil {
+		return nil, fmt.Errorf("failed serializing static pod manifest for %s to YAML: %w", client.ObjectKeyFromObject(pod), err)
+	}
+
+	return append([]extensionsv1alpha1.File{{
+		Path:        filepath.Join("/", "etc", "kubernetes", "manifests", pod.Name+".yaml"),
+		Permissions: ptr.To[uint32](0600),
+		Content:     extensionsv1alpha1.FileContent{Inline: &extensionsv1alpha1.FileContentInline{Data: staticPodYAML}},
+	}}, filesFromVolumes...), nil
+}
+
+func translateSpec(spec *corev1.PodSpec) {
+	spec.HostNetwork = true
+	spec.PriorityClassName = "system-node-critical"
+}
+
+func translateVolumes(ctx context.Context, c client.Client, pod *corev1.Pod) ([]extensionsv1alpha1.File, error) {
+	var (
+		files               []extensionsv1alpha1.File
+		addFileWithHostPath = func(hostPath, fileName, content string, desiredItems []corev1.KeyToPath) {
+			path := filepath.Join(hostPath, fileName)
+			if len(desiredItems) == 0 || slices.ContainsFunc(desiredItems, func(item corev1.KeyToPath) bool {
+				path = filepath.Join(hostPath, item.Path)
+				return fileName == item.Key
+			}) {
+				files = append(files, extensionsv1alpha1.File{
+					Path:        path,
+					Permissions: ptr.To[uint32](0600),
+					Content: extensionsv1alpha1.FileContent{Inline: &extensionsv1alpha1.FileContentInline{
```

Comment:
I'm wondering if it makes sense to encode the content in base64 for when the OSC is later also present in the cluster and not only in-memory.

Reply:
I'm open to it, but what benefits do you see when we'd do this? 👀

Reply:
Base64 allows safely encoding arbitrary data (even binary) in string representation. That's why k8s secrets use it in the first place. Hence, it would be safer to also transmit the transformed secret data in base64 as well, to ensure we can represent all valid input in the output of the translator.
Also, it would be consistent with most OSC files added for shoots, which typically use base64 as well.